### PR TITLE
Add basic iptables structures to netstack.

### PR DESCRIPTION
**Pull request by kevinGC**
Change-Id: Ib589906175a59dae315405a28f2d7f525ff8877f

### CODE REVIEW

File: pkg/tcpip/iptables/types.go [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+	Return
+)
+
+// IPTables holds all the tables for a netstack.
+type IPTables struct {
+	// mu protects the entire struct.
+	mu sync.RWMutex
+
+	// Tables maps table names to tables. User tables have arbitrary names.
+	Tables map[string]*Table
```
Comment (iangudger) [inline pkg/tcpip/iptables/types.go]: Why so many pointers?

File: pkg/tcpip/iptables/types.go [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+	Continue
+
+	// Return indicates that traversal should return to the calling chain.
+	Return
+)
+
+// IPTables holds all the tables for a netstack.
+type IPTables struct {
+	// mu protects the entire struct.
+	mu sync.RWMutex
```
Comment (iangudger) [inline pkg/tcpip/iptables/types.go]: Shouldn't this go in the parent?

Comment (kevinGC) [reply pkg/tcpip/iptables/types.go]: What parent? And reading/writing to an IPTables instance requires mu to be held.

File: pkg/tcpip/iptables/types.go (hunk 1 above)
Comment (kevinGC) [reply pkg/tcpip/iptables/types.go]: Removed the unnecessary pointers.

File: pkg/tcpip/iptables/iptables.go [hunk 3]
```diff
@@ -34,9 +34,9 @@ const (
 // all packets.
 func DefaultTables() *IPTables {
 	tables := IPTables{
```
Comment (iangudger) [inline pkg/tcpip/iptables/iptables.go]: Just return &IPTables{ ?

It looks like storing it in a named variable is no longer required.

File: pkg/tcpip/iptables/BUILD [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+go_library(
+    name = "iptables",
+    srcs = [
+        "iptables.go",
+        "targets.go",
+        "types.go",
+    ],
+    importpath = "gvisor.googlesource.com/gvisor/pkg/tcpip/iptables",
+    visibility = ["//visibility:public"],
```
Comment (iangudger) [inline pkg/tcpip/iptables/BUILD]: Shouldn't we lock this down a bit more?

File: pkg/tcpip/iptables/types.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+	Continue
+
+	// Return indicates that traversal should return to the calling chain.
+	Return
+)
+
+// IPTables holds all the tables for a netstack.
+type IPTables struct {
+	// Mu protects the entire struct.
+	Mu sync.RWMutex
```
Comment (iangudger) [inline pkg/tcpip/iptables/types.go]: Maybe just embed this field?

File: pkg/tcpip/iptables/types.go [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+	// Chains maps names to chains for both builtin and user-defined chains.
+	// Its entries point to Chains already either in BuiltinChains and
+	// UserChains, and its purpose is to make looking up tables by name
+	// fast.
+	Chains map[string]*Chain
+}
+
+// ValidHooks returns a bitmap of the builtin hooks for the given table.
+//
+// Precondition: IPTables.mu must be locked for reading.
```
Comment (iangudger) [inline pkg/tcpip/iptables/types.go]: Update. Name is out of date.

Comment (iangudger) [inline pkg/tcpip/iptables/types.go]: This feels like a leaky abstraction. Can you either let the user of this package completely handle synchronization or full encapsulate it?

**Comment by prattmic** [comment]
Merged in https://github.com/google/gvisor/commit/d60ae0ddee1799c8e45d0f7394f128ea67eb0ab8

**Pull request closed without merging**